### ISSUE TITLE
Returning TypedResult.Redirect become EmptyHttpResult

### ISSUE THREAD

**Issue by andrew-vdb** [dotnet/aspnetcore#56714]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Describe the bug

I have simple minimal api that signout and doing redirect, it always return EmptyHttpResult for no reason
It is funny because i have login endpoint with minimal api that do RedirectResult but it works

### Expected Behavior

returning TypedResult.Redirect should just work

### Steps To Reproduce

internal static async Task<RedirectHttpResult> Post(HttpContext httpContext)
{
    if (httpContext.User.Identity.IsAuthenticated)
    {
        await httpContext.SignOutAsync(CookieAuthenticationDefaults.AuthenticationScheme);
    }
    return TypedResults.Redirect("/");
}

**Comment by andrew-vdb** [comment]
Workaround (No idea why i need to do this work around, but it works)

```
internal static async Task<RedirectHttpResult> Post(HttpContext httpContext)
{
    if (httpContext.User.Identity.IsAuthenticated)
    {
        await httpContext.SignOutAsync(CookieAuthenticationDefaults.AuthenticationScheme);
        httpContext.Response.Redirect("/");
    }
    return TypedResults.Redirect("/");
}
```

**Comment by BrennanConroy** [comment]
You're returning a `Task<T>` in the first case, but because you set the return type to `Task` minimal APIs will just see `Task` and have no idea that you actually returned an `IResult` type. In the second example, you changed the return type so now minimal APIs will see your `IResult` and handle it properly.

**Comment by andrew-vdb** [comment]
Sorry, first sample will compile error, wrong copy paste

What make it work in workaround is response redirect, without that line, its not redirecting

**Comment by andrew-vdb** [comment]
Screenshot

As you can see in debugger, the delegate return RedirectHttpResult but when it goes to dynamic, it become EmptyHttpResult

<img width="521" alt="image" src="https://github.com/dotnet/aspnetcore/assets/12871379/0f07c92a-d4a6-40b3-8a57-b1170af72a55">

![image](https://github.com/dotnet/aspnetcore/assets/12871379/46e0f739-4e2b-40f5-a390-47983a3ee982)

**Comment by BrennanConroy** [comment]
Can you please share a minimal repro, because trying to reproduce this shows it working as expected.
```
app.MapPost("/p", async Task<RedirectHttpResult> (HttpContext httpContext) =>
{
    await Task.Yield();
    return TypedResults.Redirect("/");
}).AddEndpointFilter(async (c, n) =>
{
    var res = await n(c);
    return res;
});
```

**Comment by andrew-vdb** [comment]
I will try, im on holiday at this moment, i have no access to laptop for a week something

**Comment by andrew-vdb** [comment]
@BrennanConroy for some reason, that endpoint always running in this delegate, do you maybe know why the filter running on that delegate?
<img width="764" alt="image" src="https://github.com/user-attachments/assets/67d2068a-d87e-46d2-a121-10ba49acbfa4">

**Comment by BrennanConroy** [comment]
Again, please share a minimal repro, just showing screenshots of our code isn't helpful.

**Comment by andrew-vdb** [comment]
Took a while to create the repro but here you go
[AspNet.EmptyEndpointRepro.zip](https://github.com/user-attachments/files/16423691/AspNet.EmptyEndpointRepro.zip)

![image](https://github.com/user-attachments/assets/13b16273-bc75-4f08-a434-d37bb1e1034d)

**Comment by BrennanConroy** [comment]
Simplified repro:
```csharp
var g = app.MapGroup("/test").AddEndpointFilter(async (c, next) =>
{
    var res = await next(c);
    return res;
});
g.MapGet("/red", H.Redirect);

public static class H
{
    public static async Task<RedirectHttpResult> Redirect(HttpContext h)
    {
        await Task.Delay(100);
        return TypedResults.Redirect("/"); //This redirect is the cause of EmptyHttpResult
    }
}
```

This is the same as https://github.com/dotnet/aspnetcore/issues/44316 where we wrote an analyzer that should be warning you about this pattern
![image](https://github.com/user-attachments/assets/8ed71e3e-0b19-4502-9b6b-fce5dff37a20)


TLDR; if you want the `Task<T> FuncName(HttpContext)` method format, you need to cast it to `Delegate` when passing it to `Map*` methods.

**Comment by andrew-vdb** [comment]
Existing issue 2 years ago
https://github.com/dotnet/aspnetcore/issues/39956

PR that fixed it
https://github.com/dotnet/aspnetcore/pull/40235

PR that revert the fix
https://github.com/dotnet/aspnetcore/pull/42519

Another issue that discuss the revert
https://github.com/dotnet/aspnetcore/issues/45373

The reason to revert the change is to make it "reflection free"

Tbh, i do not get the "reflection free" in minimal api since AddEndpointFilterFactory use reflection, does this mean using AddEndpointFilterFactory is discouraged? (Tried to ask this in discussion but no one answer) if yes what is the alternative solution?

But
https://stackoverflow.com/questions/73426685/how-is-delegate-being-cast-to-requestdelegate-in-asp-net-core
All delegate in the end converted to RequestDelegate during "runtime"
Is this why first request very slow even using minimal api?
Then should we somehow write minimal api with RequestDelegate to avoid conversion during runtime which will make thing fast somehow

**Comment by BrennanConroy** [comment]
"reflection free" is incorrect wording. They meant non-trim safe reflection which `Map(..., Delegate)` has.

> Is this why first request very slow even using minimal api?

No, the first request is probably slow because all the generated code is created on first request.
https://github.com/dotnet/aspnetcore/issues/46313 talks about some of this.

**Comment by andrew-vdb** [comment]
Thank you for your reply, i think there is not much we can do in this issue... I will see if I can create endpoint directly with RequestDelegate as its fast start up, the only problem is that all of my filter basically broken with RequestDelegate...

https://github.com/dotnet/aspnetcore/issues/46313#issuecomment-1412998028
